src/libraries/System.IO.Compression/tests/ZipArchive/zip_UpdateTests.Comments.cs

```diff
@@ -0,0 +1,189 @@
+            string truncatedAsciiAsUtf8 = GetTruncatedString(Encoding.UTF8, largeAscii);
+            string truncatedAsciiAsLatin1 = GetTruncatedString(Encoding.Latin1, largeAscii);
+
+            char enie = '\u00d1'; // Ñ
+            string largeUnicode = new string(enie, ushort.MaxValue + 100);
+            string truncatedUnicodeAsUtf8 = GetTruncatedString(Encoding.UTF8, largeUnicode);
+            string truncatedUnicodeAsLatin1 = GetTruncatedString(Encoding.Latin1, largeUnicode);
+
+            // Detect ASCII automatically
+            yield return new object[] { null, null, string.Empty, largeAscii, truncatedAsciiAsAscii };
+            yield return new object[] { null, largeAscii, truncatedAsciiAsAscii, null, string.Empty };
+
+            // Specify encoding in constructor
+            // Preserve ASCII
+            yield return new object[] { Encoding.ASCII, null, string.Empty, largeAscii, truncatedAsciiAsAscii };
+            yield return new object[] { Encoding.ASCII, largeAscii, truncatedAsciiAsAscii, null, string.Empty };
+
+            // Force encoding to UTF8
+            yield return new object[] { Encoding.UTF8, null, string.Empty, largeAscii, truncatedAsciiAsUtf8 };
+            yield return new object[] { Encoding.UTF8, largeAscii, truncatedAsciiAsUtf8, null, string.Empty };
+
+            // Force encoding to Latin1
+            yield return new object[] { Encoding.Latin1, null, string.Empty, largeAscii, truncatedAsciiAsLatin1 };
+            yield return new object[] { Encoding.Latin1, largeAscii, truncatedAsciiAsLatin1, null, string.Empty };
+
+            // Detect unicode automatically, choose UTF8
+            yield return new object[] { null, null, string.Empty, largeUnicode, truncatedUnicodeAsUtf8 };
+            yield return new object[] { null, largeUnicode, truncatedUnicodeAsUtf8, null, string.Empty };
+
+            // Specify encoding in constructor
+            // Force encoding to UTF8
+            yield return new object[] { Encoding.UTF8, null, string.Empty, largeUnicode, truncatedUnicodeAsUtf8 };
+            yield return new object[] { Encoding.UTF8, largeUnicode, truncatedUnicodeAsUtf8, null, string.Empty };
+
+            // Force encoding to Latin1
+            yield return new object[] { Encoding.Latin1, null, string.Empty, largeUnicode, truncatedUnicodeAsLatin1 };
+            yield return new object[] { Encoding.Latin1, largeUnicode, truncatedUnicodeAsLatin1, null, string.Empty };
+        }
+
+        private static string GetTruncatedString(Encoding encoding, string largeString)
+        {
+            int bytesPerChar = encoding.GetMaxByteCount(1);
+            int encodedCharsThatFit = ushort.MaxValue / bytesPerChar;
+            int totalBytesToTruncate = encodedCharsThatFit * bytesPerChar;
+            byte[] truncatedBytes = encoding.GetBytes(largeString)[0..totalBytesToTruncate];
+            return encoding.GetString(truncatedBytes);
+        }
```

Comment:
IIRC, testing internal methods is not a very good idea, you should only test your implementation through the proper entry points (APIs). However if there is value in doing this it is better to use the same source as the one used for the product build, otherwise they might get out-of-sync and these tests will turn obsolete.

Reply:
This isn't testing an internal method, I was asked to avoid writing the exact same code as we have it internally, so there's nothing that will go out of sync.

I still need to be able to determine if a truncated string is properly truncated, so I can't think of a better way of generating one.